Add a DCC subsystem for handling the SEND/ACCEPT commands

RubenAstudillo opened the pull request:
This series of patches add at user level the following commands:
````
    /dcc          : List the DCC SEND requests
    /dcc accept n : Start downloading the n request
    /dcc resume n : Start downloading the n request appending the file on the download-dir
    /dcc cancel n : Cancel the n download
````
I've tested their implementation for a couple of weeks on a couple of fileservers. They satisfy the DCC SEND/RESUME/ACCEPT protocol for what it does a practice. What will be improved on further patches can be UI related matter for `/dcc` and see how to rate-limit the transfer so they don't consume all the bandwidth available.

The series of patches can be squash if desired.

RubenAstudillo committed:
Initial support for DCC transfers

Currently we parse a DCC event, add it to a queue on ClientState and
setup an DCCEvent that automatically starts the download on another
async thread. Bare-bone implementation. The sequence of packets sent (as
seen by wireshark) match what hexchat does on their DCC sending.

RubenAstudillo committed:
Basic user input for downloading files

RubenAstudillo committed:
Working reports on the transfers

If the Async that holds the thread finished correctly then the download
was succesful, if it throws an exception then either the user killed it
or the socket failed. Along this we are updating the counter with the
percentage downloaded. This is noticed on getEvent and deal from there.

RubenAstudillo committed:
DCC RESUME functionality

RubenAstudillo committed:
Add optional "download-dir" to configuration

It decides where to place DCC downloads, defaulting to HOME when it is
missing.

RubenAstudillo committed:
Refactor the dcc code

RubenAstudillo committed:
Handle relative paths on the download-dir option

glguy commented on src/Client/Configuration.hs:
```diff
@@ -322,10 +327,13 @@ configurationSpec = sectionsSpec "" $
              let _configDefaults = ssDefUpdate def
                  _configServers  = buildServerMap _configDefaults ssUpdates
                  _configKeyMap   = foldl (\acc f -> f acc) initialKeyMap bindings
+                 _configDownloadDir = maybe home id maybeDownloadDir
```
`fromMaybe`

glguy commented on src/Client/Configuration.hs:
```diff
@@ -228,6 +231,7 @@ loadConfiguration ::
 loadConfiguration mbPath = try $
   do (path,txt) <- readConfigurationFile mbPath
      def  <- loadDefaultServerSettings
+     home <- getEnv "HOME" -- guarranted by POSIX?
```
Let's use [getHomeDirectory](http://hackage.haskell.org/package/directory-1.3.3.2/docs/System-Directory.html#v:getHomeDirectory)

glguy commented on src/Client/Configuration.hs:
```diff
@@ -322,10 +327,13 @@ configurationSpec = sectionsSpec "" $
                                "Initial setting for visibility of ping times"
-     return (\def ->
+     maybeDownloadDir       <- optSection' "download-dir" stringSpec
+                               "Path to the directory where to place DCC downloads"
```
`"Path to DCC download directory. Defaults to home directory."`

glguy commented on src/Client/EventLoop.hs:
```diff
@@ -285,8 +291,11 @@ doNetworkLine networkId time line st =
+                    let (replies, dccUp, st3) =
+                          applyMessageToClientState time irc networkId cs st2
 
+                    maybe (return ())
```
Instead of `maybe (return ())`: [traverse_](http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#v:traverse_)

glguy commented on src/Client/State/Focus.hs:
```diff
@@ -45,6 +45,7 @@ data Subfocus
   = FocusMessages    -- ^ Show messages
   | FocusInfo        -- ^ Show channel metadata
   | FocusUsers       -- ^ Show channel user list
+  | FocusDCC
```
It's obvious, but could you put a haddock command on this constructor?

RubenAstudillo replied:
I forgot!

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+-- | Status of a connection at certain @Key@
+data ConnectionStatus
+  = CorrectlyFinished | UserKilled | LostConnection | Downloading
+  | Pending | NotExist
```
Is it necessary to have `NotExist`? Could we use `Maybe ConnectionStatus` for cases where nonexistence is possible? This would save an impossible case in the View module.

RubenAstudillo replied:
`NotExist` is neccesary for the case `cmdDcc` is called with an invalid key. It's never stored on the `ConnectionStatus` of a `DCCTransfer`, that is why it is impossible to be shown on the View module. That also why a `Maybe ConnectionStatus` won't work as when we set that field, it is never `NotExist`.

See also the `statusAtKey` function, that is the only one capable of returning a `NotExist`.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+                    -- user killed, thrown by async >= 2.2
+                    , E.Handler (\(_ :: AsyncCancelled) ->
+                                   return (UserInterrupted key))
+                    -- Compatibility hack
```
glirc doesn't need compatibility, let's just bump the lower-bound on async and knock out hacks

RubenAstudillo replied:
I was gonna ask this! Got it.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+                                    wait outThread
+  where
+    param = ConnectionParams
+              { cpFamily = AF_INET, cpHost = from
```
Do we have to back IPv4 in here? What would it take to work on both 4 and 6?

RubenAstudillo replied:
[Protocol stuff](http://www.irchelp.org/protocol/dccspec.html) (Yeah, this old stuff is the authority here). The DCC that uses IPv6 is implementation specific. I found a fileserver that uses IPv6 and I am asking for the fileserver code plus checking what hexchat does with IPv6. I will do some tests and send a patch after that.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+  let offerList = I.toDescList (_dsOffers state)
+      predicate fileName (key, offer)=
+          (_dccFileName offer == fileName)
+          && userFrom == (_dccFromInfo offer)
```
Extra parentheses not needed

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+             <*> (ipv4Dotted <$> (space *> decimal))
+             <*> (space *> decimal)
+             <*> (space *> decimal)
+
```
I'd rather just see do-notation than a nested applicative expression

```haskell
sendFormat network userFrom =
  do name      <- Text.unpack <$ string "SEND" <* space <*> nameFormat
     addr      <- ipv4Dotted <$ space <*> decimal
     port      <- space *> decimal
     totalsize <- space *> decimal
     return (DCCOffer network userFrom addr port name totalsize 0)
```

RubenAstudillo replied:
I'll also change `acceptFormat`.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+
+      txt = "RESUME" <> " " <> quoting <> filename <> quoting <> " " <> port
+            <> " " <> sizeoffset'
+      target = Text.unpack $ views uiNick idText (_dccFromInfo offer)
```
`views (uiNick . dccFromInfo) (Text.unpack . idText) offer`

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+acceptUpdate :: DCCUpdate -> DCCState -> DCCState
+acceptUpdate (Accept k port offset) state =
+  case view (dsOffers . at k) state of
+    Nothing -> error "Not a valid key"
```
Could this kill the client if something went wrong? Could we silently ignore a missing key instead? What would the consequences be?

RubenAstudillo replied:
We could simply ignore it. This function came to be as some code that was too big to be included directly on `doDCCUpdate`, so the checking that that `case` was not `Nothing` it's done there (and it's the only place where `acceptUpdate` is called). I put an `error` so playing on ghci would show that something was obviously wrong, but I will change that to "do nothing".

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+--   on the supervisor thread at @Key@
+reportStopWithStatus :: Key -> ConnectionStatus -> DCCState -> DCCState
+reportStopWithStatus key newstatus dccstate =
+  let transfer = maybe (error "Wrong index") id $ view (dsTransfers . at key) dccstate
```
What are the consequences of `error` here? When can this happen? Can we handle this more gracefully or silently?

RubenAstudillo replied:
The `key` is guaranteed to be valid as is generated by the thread at that key and we don't remove any item of `DCCState`. But I didn't know how to convince the compiler that prism branch chosen was the valid one. I'll just match on the `Just` value returned.

RubenAstudillo replied:
There is a single call to `reportStopWithStatus` and it guaranteed to be called on a valid `key`. I will just match on the `Just` returned by the `prism`.

glguy commented on src/Client/Commands.hs:
```diff
@@ -929,6 +932,16 @@ commandsList =
+       \  /dcc accept #n : start downloading the #n pending offer\n\
+       \  /dcc resume #n : same as accept but appending to the file on `download-dir`\n\
+       \  /dcc cancel #n : cancel the download #n \n\n"
+    $ ClientCommand cmdDcc noClientTab
```
Future work: smarter tab completion for the commands (accept, resume, cancel) and for the active DCC IDs. Not necessary for this PR however.

RubenAstudillo replied:
Right! it would be nice!

glguy reviewed (changes requested):
I think you've done a nice job of matching the style and structure of the client with this change. I'd got some things annotated below that we should look at.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+  , _dccSize     :: Word32 -- ^ Size of the whole file, per protocol
+                           --   restricted to 32-bits
+  , _dccOffset   :: Word32 -- ^ Byte from where the transmission starts
+  } deriving (Show, Eq)
```
Let's make the fields strict where it makes sense, basically everything but the fields represented by a `[]`

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+  mempty = DCCState mempty mempty
+
+instance Semigroup DCCState where
+  (DCCState off1 tran1) <> (DCCState off2 tran2) =
```
We don't need parentheses around these patterns

RubenAstudillo replied:
Removing all together

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+instance Monoid DCCState where
+  mempty = DCCState mempty mempty
+
+instance Semigroup DCCState where
```
I'd prefer just not having this instance. I find reusing `Monoid` and `Semigroup` for things like this to be a hack. If we keep it, let's put a haddock command explaining the behavior.

RubenAstudillo replied:
Yeah, I only provide `Semigroup` for `Monoid`. I wanted to use `empty` as a default value for `ClientState`. I will provide just a constant as alternative.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+  { _dtThread   :: Maybe (Async ()) -- ^ If Nothing, the thread was killed
+                                       --   and stopped.
+  , _dtProgress :: Word32 -- ^ Percentage of progress
+  , _dtStatus   :: ConnectionStatus
```
Strict fields when it makes sense.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+  | SocketInterrupted Key
+  | UserInterrupted Key
+  | Accept Key PortNumber Word32 -- update that DCC RESUME triggers
+  deriving (Show, Eq)            -- Word32 is the offset
```
Strict fields, small fields like these will unpack automatically and we don't benefit from any extra laziness here.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+      do -- Has to decouple @send@ from @recv@, tells how much
+         -- have we downloaded.
+         recvChan1 <- atomically newTChan
+         recvChan2 <- atomically $ dupTChan recvChan1
```
Could you help me understand why we had to duplicate the channel? I just don't quite understand the thread model here.

RubenAstudillo replied:
OK, so `recvChan1` is what `sendStream` uses for acknowledgement of the `recv`, this sends how many bytes have we received from that `socket` so far. DCC SEND/ACCEPT as a protocol says that after each `recv` you have to `send` that acknowledgement. In practice the server doesn't care about the acknowledgement, it will keep sending you stuff and not let you `send` your acknowledgement until the transfer finishes (which is why we start an extra thread to be blocked on these `send` request). I've checked this behavior with wireshark on at least two machines and with different fileservers on rizon, it fails if you do what the bare protocol describes.

But we also want to report how much progress have we done on a download, but maintaining the blocking behavior of `sendStream` has respect `recvChan1`. The solution is `dupTChan recvChan1` onto `recvChan2`, so `report` blocks on the latter generating `DCCUpdate` event that the `eventLoop` process and lets `sendStream` have the blocking behavior respect to `recvChan1`.

glguy commented on src/Client/State/DCC.hs:
```diff
@@ -0,0 +1,348 @@
+                   compareAndUpdate curPercent
+
+percent :: Word32 -> Word32 -> Word32
+percent a total = fromIntegral $
```
```haskell
percent a total = fromIntegral (fromIntegral a * 100 `div` fromIntegral total :: Word64) -- avoid overflow
```

RubenAstudillo committed:
Fix/add the comments for the branch

Most of this stuff was annotated on
https://github.com/glguy/irc-core/pull/64 .

RubenAstudillo committed:
Correctly acknowledge the size on DCC RESUME

We usually send the received data on the socket, but on the case of a
resume we ought to start acknowledging the received data plus the
starting offset. This is not on the protocol but match what the server
expect on practice.

glguy commented:
I tried out the transfer and had some issues and questions:

1. When I ran /dcc with no active transfers the window was empty. I've been showing a message for empty lists. Try `/masks q` on a channel with no quiet modes.
2. My first transfer ended with: Socket failure 74% (transfer from Textual). I don't know what went wrong, but the file only partially transferred. Which clients have you tried with?
3. Incoming DCC offers not rendered in client, we should add a new message type similar to CtcpAction so that the user won't have to use the detail view to learn about DCC offers
4. How do I remove failed transfers from /dcc? After my transfer above I couldn't figure out how to remove the transfer from the list.
5. DCC filenames are not sanitized; they could overwrite existing files, and they could contain directory separators

glguy reviewed (changes requested):
I tried out the transfer and had some issues and questions:

1. When I ran /dcc with no active transfers the window was empty. I've been showing a message for empty lists. Try `/masks q` on a channel with no quiet modes.
2. My first transfer ended with: Socket failure 74% (transfer from Textual). I don't know what went wrong, but the file only partially transferred. Which clients have you tried with?
3. Incoming DCC offers not rendered in client, we should add a new message type similar to CtcpAction so that the user won't have to use the detail view to learn about DCC offers
4. How do I remove failed transfers from /dcc? After my transfer above I couldn't figure out how to remove the transfer from the list.
5. DCC filenames are not sanitized; they could overwrite existing files, and they could contain directory separators

RubenAstudillo committed:
Sanitization of filepath for DCC transfers

At parse time, we remove path separators from the name and present just
that. When the user issues a "/dcc (accept|resume)" we check the
download directory for a file of the same name, if it is a directory we
fail, otherwise we assume that if the user said "resume" we want to
append to this file and if he said "accept" we overwrite.

RubenAstudillo committed:
Basic count of offers on the DCC subfocus

Added a basic countdown of what is on the list. It's displayed even on a
empty list.

RubenAstudillo commented:
I'll answer on order

1. I added a basic countdown based on the code for `MaskList.hs` that shows at the bottom the total and how many transfers are running.

2. That's odd, at least initiating a transfer from hexchat to this branch of code works as expected. Also file-servers on Rizon work too. I've had problems trying to run `Textual`. We could try to do a transfer from your end to my client and see what I log.

3. Yes, pressing F2 to see if something has come will get old. I wanna read the source code a little more to see how the /view-* functionality is implemented.

4. Currently you can't remove anything from the list. There is nothing fundamental stopping it, I just didn't want to complicate the options under "/dcc" with an hypothetical "/dcc remove".  I am thinking on possible directions on how to present all the functionality to the user and not being overbearing. 

5. Now they are. At parse time we use `takeFileName` to clear path separators and at "/dcc (resume|accept)" we check if the file is a directory on the download-dir. We assume that if the user says "resume" we want to append to whatever file is on the download-dir with the same name and if he says "accept" he wants to overwrite.

edit: Most of the commits appear new because of a rebase, just the last 2 are related to this answer

RubenAstudillo committed:
Document the DCC features on README.md

RubenAstudillo committed:
Implement /dcc clear command to remove from dcc offer list

RubenAstudillo committed:
Check download-dir is an actual directory and writeable

When it doesn't satisfy the conditions, we throwIO an
ConfigurationMalformed exception.

RubenAstudillo commented:
I added a "/dcc clear" command that removes transfers from the "/dcc" list (if they are allowed to be removed).

RubenAstudillo committed:
show a WindowLine when receiving a DCC SEND message

When constructing a IrcSummary, we treat a "DCC SEND" message as their
own constructor. This is so a WindowLine will be shown on them but the
rest of the "DCC" commands like "RESUME" and "ACCEPT" are treated like
before.

RubenAstudillo commented:
Now it signals as a `WindowLine` when a "DCC SEND" command is received (so the user checks the "/dcc" list). It looks like this

![gato](https://user-images.githubusercontent.com/1875145/53385737-3aef9680-395e-11e9-8c41-bd323f1d1296.png)
The previous sequence shows only the "DCC SEND", even if send a "DCC RESUME" and it's followed by a "DCC ACCEPT" those are not showed, as it's just an implementation detail on how resume is done on DCC.

![screenshot_20190226_003157](https://user-images.githubusercontent.com/1875145/53385775-62466380-395e-11e9-9559-d9c660a3a959.png)
This is corresponding view on "/dcc".

With this, I consider the pull request is completed :-)
